internal/counter/counter.go

```diff
@@ -0,0 +1,230 @@
+/*
+Copyright (c) 2024 Diagrid Inc.
+Licensed under the MIT License.
+*/
+
+package counter
+
+import (
+	"context"
+	"time"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/diagridio/go-etcd-cron/api"
+	"github.com/diagridio/go-etcd-cron/internal/client"
+	"github.com/diagridio/go-etcd-cron/internal/garbage"
+	"github.com/diagridio/go-etcd-cron/internal/grave"
+	"github.com/diagridio/go-etcd-cron/internal/key"
+	"github.com/diagridio/go-etcd-cron/internal/scheduler"
+)
+
+// Options are the options for creating a new counter.
+type Options struct {
+	// Name is the name of the job to count.
+	Name string
+
+	// Key handles the key to use writing to etcd.
+	Key *key.Key
+
+	// Client is the etcd client to use.
+	Client clientv3.KV
+
+	// Schedule is the schedule of this job.
+	Schedule scheduler.Interface
+
+	// Job is the job to count.
+	Job *api.JobStored
+
+	// Yard is a graveyard for signalling that a job has just been deleted and
+	// therefore it's Delete informer event should be ignored.
```

Comment:
Is this correct - bc the delete informer event is the in mem 'event'? and the `yard` is the db 'absolute truth'?

Reply:
Yes. Consider the case where a scheduler Delete's a Job as it has expired. This means that we _don't enqueue the job again and delete the job key from ETCD. Since we inform on job delete events, we need to track recently deleted jobs so that we don't force the scheduler to also `Dequeue` a job every time a job expires unnecessarily.